### ISSUE TITLE
Large memory consumption of MessageIdProvider

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#323]
**The issue:**
I tested the 2.0.0M3 in production environment and we have a bug issue with memory consumption.
By analyzing the heap dump. I saw main consumption comes from the `InMemoryMessageIdProvider` and in particular all the `MessageIdTracker`.

**The way it is:**
`InMemoryMessageIdProvider` uses a LeastRecentlyUsedCache with 150 000 as max size by default.  This map is never emptied.
Each entry of this LRU cache contains a `MessageIdTracker` with a Map initialized at 65536.

**My approximation was for 150 000:**
150 000 x 65536 x 8 bytes/element (for an empty map in the tracker) = 78.64 GB
150 000 x 65536 x 8+4+8 (ref+int+long) bytes/element (for an full map ) = 196.60 GB
The dump file seems to confirm those estimations, as for 2415 peers (where maps is  almost empty) we consume 1.27GB ~= 1.26GB ~= 2415×65536×8. (The estimation is is even so closed, that it makess me feel I maybe made an error ?, but I suppose the order of magnitude is here)

**Knowing that :**
- a MessageID should NOT be reused in **EXCHANGE_LIFETIME** (247s by default).
- we are now able to track MessageID by peer (great improvement)
- MessageID generation concerns only outgoing Message(Request, separated response).
That's mean "using a simple rotation for MessageID generation" you should send more than 265 messages by seconds during more than 247s (TOTAL_NO_OF_MIDS/TOTAL_NO_OF_MIDS) to get collision.

**My point of view :**
As I understand the code the "big" Map in `MessageIdTracker` is only used to be able to know if we will reuse an ID or not and so log a warning.
A counter and increment it with random initialization seems enough to me. We don't need the map in MessageIdTracker. If there is users  who think there will maybe send more than 265 msg/s and need warning for this. they are free to implement their own InMemoryMessageIdProvider.

**Comment by boaks** [comment]
so have a look at #271

**Comment by boaks** [comment]
Just to say:
If a MID collision happens (too early reuse), it's very hard to understand, what's going on.
(see #30).

Therefore a implementation, which doesn't consume that much memory, but reports a "too early reuse" would be preferred by me.

**Comment by sophokles73** [comment]
I like the idea behind #271 :-)

**Comment by sbernard31** [comment]
I'm looking at your code. I was thinking of a kind of optimization like that, but after reflexion I think that is clearly not needed.

I understand  when MID was shared by all the peers, collision of ID was a problem, but now I think this should not happen anymore. More than 265 msg/s in parallel, which kind of use case this can be ?

**Comment by boaks** [comment]
Even if its no "CoAP" use case, to have a warning instead of "its not working" is simply better. We can make the number of groups in #271 configurable, so even with 16 groups it's better then nothing.

**Comment by sophokles73** [comment]
I tend to agree with @boaks in that I believe it is better to know that we are running out of MIDs than wondering, why messages are discarded as duplicates (if we are able to detect that at all). Using a small number of groups (even down to just 2 would be beneficial, I guess) should come close to just using a single counter per peer, don't you think @sbernard31 ?

**Comment by sbernard31** [comment]
FMPOV having "complex code" for "just logging a warning in an extremely hypothetical use case"  is not a good tradeoff.
We can have several implementations in californium, so we just need to agree on the default one. Personally I prefer the simplest one for the reason I exposed before, but if we change the current one for another which don't consume several GB of memory it's ok for me. 

In Leshan, we will probably go for the simplest one by default.

**Comment by boaks** [comment]
Sometime its worth to have a warning, just to ensure, that it's never shown :-).

So I go for make #271 more configurable and leshan can replace it by a even simpler one.

**Comment by boaks** [comment]
> In Leshan, we will probably go for the simplest one by default.

So I hope, that the leshan user can decide, which one they want to use :-).

**Comment by sbernard31** [comment]
Currently there is no easy way to change it. 

For now, There is a setter for  MessageIDProvider in MessageExchangeStore. But the only way to access to MessageExchangeStore is to create a new one, keep a reference on it and use it for the `CoapEndpoint` constructor.

a setter(usable before we start the endpoint)  and a getter on MessageExchangeStore could help.

**Comment by boaks** [comment]
After trying to use #271 I recognized, there is more to do :-).
I hope I can finish an new PR which enables to select between id tracker implementations using the NetworkConfig.

**Comment by boaks** [comment]
See #328

**Comment by sbernard31** [comment]
I'm not sure this warning worth as much effort.
Personally, I'm out on this one ;)
If you are interested, there is a simple and working implementation of MessageIDProvider  [here](https://github.com/eclipse/leshan/pull/355/commits/c0ef06123d85b2480c7711265c484d007d5bb3df#diff-e486af47d4c35e8a2ee71036502eac08R26).

**Comment by boaks** [comment]
> Personally, I'm out on this one ;)

So this issue is to be closed?

**Comment by sbernard31** [comment]
>So this issue is to be closed?

I don't think so. The problem described is still here. We should close it when a new implementation will be integrated.

Personally, I would go for integrating the Manu's implementation,  close this issue and move on.
But if you both think that the others one are worthy, I'm just saying that I let you polish that :)

**Comment by boaks** [comment]
@sbernard31 
Does PR #328 fix this leak? If so, could the issue be closed?

**Comment by sbernard31** [comment]
We can close it :)
